src/routing/bootstrap.rs

```diff
@@ -0,0 +1,653 @@
+    ) -> Result<(Node, Section)> {
+        let (elders_info, section_key) = self
+            .bootstrap(bootstrap_addrs, relocate_details.as_ref())
+            .await?;
+
+        let relocate_payload = if let Some(details) = relocate_details {
+            Some(self.process_relocation(&elders_info, details)?)
+        } else {
+            None
+        };
+
+        self.join(elders_info, section_key, relocate_payload).await
+    }
+
+    // Send a `BootstrapRequest` and waits for the response. If the response is `Rebootstrap`,
+    // repeat with the new set of contacts. If it is `Join`, proceeed to the `join` phase.
+    async fn bootstrap(
+        &mut self,
+        mut bootstrap_addrs: Vec<SocketAddr>,
+        relocate_details: Option<&SignedRelocateDetails>,
+    ) -> Result<(EldersInfo, bls::PublicKey)> {
+        loop {
+            self.send_bootstrap_request(mem::take(&mut bootstrap_addrs), relocate_details)
+                .await?;
+
+            let (response, sender) = self.receive_bootstrap_response().await?;
+
+            match response {
+                BootstrapResponse::Join {
+                    elders_info,
+                    section_key,
+                } => {
+                    info!(
+                        "{} Joining a section {:?} (given by {:?})",
+                        self.node, elders_info, sender
+                    );
+                    return Ok((elders_info, section_key));
+                }
+                BootstrapResponse::Rebootstrap(new_bootstrap_addrs) => {
+                    info!(
+                        "{} Bootstrapping redirected to another set of peers: {:?}",
+                        self.node, new_bootstrap_addrs,
+                    );
+                    bootstrap_addrs = new_bootstrap_addrs.to_vec();
+                }
+            }
+        }
+    }
+
+    async fn send_bootstrap_request(
+        &mut self,
+        recipients: Vec<SocketAddr>,
+        relocate_details: Option<&SignedRelocateDetails>,
+    ) -> Result<()> {
+        let destination = match relocate_details {
+            Some(details) => *details.destination(),
+            None => self.node.name(),
+        };
+
+        let message = Message::single_src(
+            &self.node.keypair,
+            self.node.age,
+            DstLocation::Direct,
+            Variant::BootstrapRequest(destination),
+            None,
+            None,
+        )?;
+
+        debug!("{} Sending BootstrapRequest to {:?}", self.node, recipients);
+
+        let _ = self.send_tx.send((message.to_bytes(), recipients)).await;
+
+        Ok(())
+    }
+
+    async fn receive_bootstrap_response(&mut self) -> Result<(BootstrapResponse, SocketAddr)> {
+        while let Some((message, sender)) = self.recv_rx.recv().await {
+            match message.variant() {
+                Variant::BootstrapResponse(response) => {
+                    if !self.verify_message(&message, None) {
+                        continue;
```

Comment:
could be an issue of block if the sender doesn't send back response anymore?
and same issue for the wild case below

Reply:
Same issue